### PR TITLE
Added function pointer support

### LINKED ISSUE #111: Add support for function pointers
I will try to implement this.

### PR DESCRIPTION
**Pull request by initram**
Fixes #111

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -608,6 +609,36 @@ module.exports = grammar({
...
+
+    unmanaged_calling_convention: $ => choice(
+      'Cdecl',
+      'Stdcall',
+      'Thiscall',
+      'Fastcall',
+      commaSep1($.identifier)
+    ),
+
+    funcptr_parameter: $ => seq(
```
Comment (damieng) [inline grammar.js:637]: Roslyn calls this `FunctionPointerParameter` so perhaps `function_pointer_parameter` ?

File: grammar.js [hunk 2]
```diff
@@ -608,6 +609,36 @@ module.exports = grammar({
...
+    function_pointer_type: $ => seq(
+      'delegate',
+      '*',
+      optional($.calling_convention_specifier),
+      '<',
+      commaSep1($.funcptr_parameter),
+      '>'
+    ),
+
+    calling_convention_specifier: $ => choice(
```
Comment (damieng) [inline grammar.js:621]: Roslyn calls this `FunctionPointerCallingConvention` so perhaps `function_pointer_calling_convention` ?

File: grammar.js [hunk 3]
```diff
@@ -608,6 +609,36 @@ module.exports = grammar({
...
+      'managed',
+      seq(
+        'unmanaged',
+        optional(seq('[', $.unmanaged_calling_convention, ']'))
+      )
+    ),
+
+    unmanaged_calling_convention: $ => choice(
```
Comment (damieng) [inline grammar.js:629]: Roslyn definitely seems to vary from the proposal grammar on this one. They have line 625 as a rule called `FunctionPointerUnamangedCallingConventionList` and then each convention inside is a `FunctionPointerUnmanagedCallingConvention`.

File: grammar.js [hunk 4]
```diff
@@ -608,6 +609,36 @@ module.exports = grammar({
...
+      'Cdecl',
+      'Stdcall',
+      'Thiscall',
+      'Fastcall',
+      commaSep1($.identifier)
+    ),
+
+    funcptr_parameter: $ => seq(
+      repeat(choice('ref', 'out', 'in', 'readonly')),
```
Comment (damieng) [inline grammar.js:638]: It can only be one of these, perhaps add `ref readonly` as the final option instead?

**Review by damieng — CHANGES_REQUESTED** [review]

File: grammar.js (hunk 1 above)
Comment (damieng) [reply grammar.js:637]: Also think we're missing corpus coverage for this.

**Comment by initram** [comment]
I have now made changes to align it more with the AST produced by roslyn. Let me know if there are any other feedback.

**Review by damieng — APPROVED** [review]

**Comment by damieng** [comment]
Looks great, thanks!

**Pull request merged**